File: hu_api/app/models/hotel.rb
class Hotel < ActiveRecord::Base
  belongs_to :city
  has_many :availabilities

  validates :city_id, presence: true
  validates :name, presence: true, uniqueness: {scope: :city_id}

  scope :by_city_or_hotel_name, ->(name) do
    name = "%#{name}%".upcase
    joins(:city).where('cities.name LIKE UPPER(?) OR hotels.name LIKE UPPER(?)',name,name)
  end

  scope :by_id, ->(id){ eager_load(:city).where(id: id) }

  scope :by_city_id, ->(city_id){ eager_load(:city).where(city_id: city_id) }

  scope :available_from, ->(start_date){ joins(:availabilities).where('availabilities.day >= ?', start_date)}

  scope :available_until, ->(end_date){ joins(:availabilities).where('availabilities.day <= ?', end_date)}

  scope :available_from_until, ->(start_date, end_date) do
    available_from(start_date.to_date).
        available_until(end_date.to_date).
        group('hotels.id').
        having('COUNT(*) >= ?', elapsed_days(start_date, end_date))
  end

  private

  def self.elapsed_days(start_date, end_date)
    (start_date.to_date..end_date.to_date).count
  end

end

File: interface/app/scripts/services/Hotel.js
angular.
  module('huHotelSearchApp').
  service("Hotel", function($http, City){
    var self = this;

    self.getHotels = function (location) {
      return $http({
        method: "get",
        url: "http://localhost:3000/api/hotels.json",
        params: {by_city_or_hotel_name: location}
      });
    };

    self.getHotelsByParams = function(params){
      var url = "http://localhost:3000/api/hotels.json?";

      if (params.id) {
        url += 'by_id=' + params.id;
      }
      if (params.city_id) {
        url += 'by_city_id=' + params.city_id;
      }

      if (params.start_date && params.end_date){
        url += '&available_from_until[start_date]='+ params.start_date;
        url += '&available_from_until[end_date]='+ params.end_date;
      }
      return $http({
        method: "GET",
        url: url });
    }
  });
